[TRTLLM-4983] feat: enable overlap scheduler between draft forwards

lfr-0531 opened the pull request:
## Description

- Enable the overlap scheduler between different draft forwards.
- Add a new Eagle3ResourceManager to manage the hidden states, and remove the extra model input.
- Move the eagle3 fc into the model forward.
- Move the h2d copy to the end of the _prepare_tp_inputs to hide the CPU time.
- Disable CUDA graph for the 1st draft forward.

I'll collect more performance and accuracy data and update the results here.

lfr-0531 committed:
add overlap scheduler to multiple draft engine forward (1st commit).

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
add a simplified prepare_inputs for draft model engine.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
remove more h2d in attm_metadata.prepare.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix eagle hidden state/cuda graph/hidden states gather issues.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix metadata.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
add eagle3 resource manager to manage hidden states.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
fix eagle3 resource manager and force disable cuda graph for the 1st draft forward.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
move the h2d copy to the end of the _prepare_tp_inputs.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #7897](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/7897/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #5706](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/5706/) completed with status: 'FAILURE'

mikeiovine reviewed (commented):
Shall we measure the performance gain? Since our initial goal is to hit parity with vLLM, I think using Llama 3.3 70B on Hopper makes sense. I got these numbers on 8xH200 today.

| Framework | Max Batch Size | OSL | Eagle (draft len = 3)? | Output tok/sec |
| --- | --- | --- | --- | --- |
| TRTLLM | 8 | 256 | No | 876 |
| TRTLLM | 8 | 256 | Yes | 1154 |
| vLLM | 8 | 256 | No | 715 |
| vLLM | 8 | 256 | Yes | 1430 |

A note about the dataset: I used gsm8k from [here](https://github.com/SafeAILab/EAGLE/blob/main/eagle/data/gsm8k/question.jsonl). For the llama 3.3 eagle drafters we have from the paper author's, **you have to apply the tokenizer's chat template** or the AR will drop significantly and performance will be regressed. I will send you a preprocessed dataset that is compatible with `trtllm-bench`.

lfr-0531 commented on tensorrt_llm/_torch/attention_backend/interface.py:
```diff
@@ -257,6 +256,11 @@ def prepare(self):
         Hook to be called before the forward step of the model.
         """
 
+    def prepare_device(self):
+        """
+        Hook to be called before the forward step of the model.
+        """
+
```
I want to put all those H2D copies at the end of the `prepare_inputs` to overlap as many CPU operations as possible. Here, the `prepare_device` is used for those H2D in the metadata.

For example, for the `input_ids`, the `self.input_ids_cuda.copy_(new_tokens_device, non_blocking=True)` can't be executed until we get the `new_tokens_device`. So the CPU operations after this H2D will be blocked.

I'll do more perf evaluations to see if this change in metadata can help or not.

lfr-0531 commented on tensorrt_llm/_torch/models/modeling_llama.py:
```diff
@@ -1273,6 +1271,9 @@ def forward(
         hidden_states: Optional[torch.Tensor] = None,
         **kwargs,
     ) -> torch.Tensor:
+        if hidden_states is None:
+            hidden_states = spec_metadata.get_hidden_states(
+                preprocess_func=self.apply_eagle3_fc)
```
Two reasons for this change:
- We want this FC to also be captured by the CUDA graph. Otherwise, there will be large bubbles around this GEMM kernel.
- Make the code clean. We don't need to run some layers in the executor and move all of them to the model forward.

For the CUDA graph, you're right, with this change, we cannot enable CUDA graph for the 1st draft forward. My current thought is to set the sequence length of the 1st draft forward to be `1+max_draft_len`, similar to the padding in https://github.com/NVIDIA/TensorRT-LLM/pull/4773. When capturing CUDA graphs for the draft model engine, we can capture two CUDA graphs for each batch size: one with `seq_len=1` and another with `seq_len=1+max_draft_len`. The FC can be captured in the cuda graph with `seq_len=1+max_draft_len`, and we can use this cuda graph for the 1st draft forward, and another one for the subsequent draft forwards.

lfr-0531 replied:
Agree, especially for the 1st draft forward.

lfr-0531 committed:
revert the metadata prepare changes.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #8377](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8377/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8377](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8377/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `5267115`

tensorrt-cicd commented:
[PR_Github #8385](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8385/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8385](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8385/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6076](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6076/) completed with status: 'FAILURE'

lfr-0531 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #8444](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8444/) [ run ] triggered by Bot

mikeiovine reviewed (commented):
Finished reviewing. The changes to enable the overlap scheduler look good to me. I think it's also OK to skip CUDA graphs for the first iteration for now - we can try to embrace the dynamic nature of EAGLE3 and experiment with piecewise CUDA graphs via torch.compile instead. Given that EAGLE2 is in the works and that one doesn't have this FC problem, I'm not too worried about it any more.

mikeiovine commented:
By the way, I'm testing on trtllm-bench and I'm hitting the following key error:
```
  File "/code/tensorrt_llm/tensorrt_llm/_torch/pyexecutor/py_executor.py", line 1878, in _prepare_draf
t_tokens
    outputs = self.draft_model_engine.forward(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_con
text
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/code/tensorrt_llm/tensorrt_llm/_torch/utils.py", line 67, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/tensorrt_llm/tensorrt_llm/_torch/pyexecutor/model_engine.py", line 2059, in forward
    inputs, gather_ids = self._prepare_inputs(scheduled_requests,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/nvtx/nvtx.py", line 122, in inner
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/code/tensorrt_llm/tensorrt_llm/_torch/pyexecutor/model_engine.py", line 1981, in _prepare_inp
uts
    return self._prepare_tp_inputs(scheduled_requests, kv_cache_manager,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/code/tensorrt_llm/tensorrt_llm/_torch/pyexecutor/model_engine.py", line 1372, in _prepare_tp_
inputs
    spec_metadata.prepare()
  File "/code/tensorrt_llm/tensorrt_llm/_torch/speculative/eagle3.py", line 159, in prepare
    start_idx = self.eagle3_resource_manager.start_indices[slot_id]
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
KeyError: None
[06/11/2025-19:45:46] [TRT-LLM] [E] Encountered an error in decode: None
```

mikeiovine commented:
The above error only occurs when CUDA graph padding is on - looks like the bad request ID comes from a dummy request. I was able to measure **669 output TPS** after disabling padding with the same settings as the previous benchmark. The AR actually looks good, so I'm not sure what's causing the regression. Perhaps the CUDA graphs for the first iteration are actually necessary?

lfr-0531 commented:
> The above error only occurs when CUDA graph padding is on - looks like the bad request ID comes from a dummy request. ~I was able to measure **669 output TPS** after disabling padding with the same settings as the previous benchmark. The AR actually looks good, so I'm not sure what's causing the regression. Perhaps the CUDA graphs for the first iteration are actually necessary?~
> 
> The regressed TPS was due to a misconfiguration on my side. The real TPS with this PR is **1131**; much closer but still just barely slower than our baseline on `main` (1154).

Thanks,  Mike! I can also reproduce the similar perf results:
| Batch Size | Max Draft Len | Branch                | Total Output Throughput (tokens/sec) | Acceptance Rate |
| ---------- | ------------- | --------------------- | ------------------------------------ | --------------- |
| 1          | 0             | main ToT              | 56.0135                              | 1.00            |
| 1          | 0             | Eagle3 overlap branch | 55.9898                              | 1.00            |
| 1          | 3             | main ToT              | 97.8856                              | 2.83            |
| 1          | 3             | Eagle3 overlap branch | 94.3258                              | 2.83            |
| 8          | 0             | main ToT              | 390.0602                             | 1.00            |
| 8          | 0             | Eagle3 overlap branch | 395.3684                             | 1.00            |
| 8          | 3             | main ToT              | 520.9109                             | 2.81            |
| 8          | 3             | Eagle3 overlap branch | 516.4105                             | 2.81            |

My cmd:
```
TRTLLM_ENABLE_PDL=1 trtllm-bench \
	--model ${PATH_TO_Llama-3.3-70B-Instruct} \
	throughput \
	--dataset ./gsm8k_template_256.json \
	--backend pytorch \
	--num_requests 80 \
	--concurrency 1 \
	--max_batch_size 1 \
	--tp 8 \
        --extra_llm_api_options ./extra-llm-api-config.yml
```
extra-llm-api-config.yml:
```
enable_attention_dp: false
disable_overlap_scheduler: true
use_cuda_graph: true
cuda_graph_padding_enabled: false
print_iter_log: false
speculative_config:
    decoding_type: Eagle
    max_draft_len: 3
    pytorch_eagle_weights_path: ${PATH_TO_EAGLE3-LLaMA3.3-Instruct-70B}
    eagle3_one_model: false
kv_cache_config:
    enable_block_reuse: false
    free_gpu_memory_fraction: 0.7
```

I'm comparing the nsys profiles and trying to find the root cause.

lfr-0531 commented:
From the nsys, we got some good cases:
![good_case](https://github.com/user-attachments/assets/9a4958c2-51cb-402a-8ed4-17ea890283f4)

And some bad cases:
![bad_case](https://github.com/user-attachments/assets/7c59cafe-5138-4c9b-89eb-97d07d4c38ec)

It is because the kernel time < prepare_inputs time. 

At the beginning, I added a simplified `_prepare_draft_tp_inputs`, but it felt like the original `_prepare_tp_inputs` would work, so I removed it. Now it seems we have to add it back.

For such a llama3.3 70B case with BS=1, the draft model forward only takes about 300us. But the `_prepare_tp_inputs` needs to use around 900us (host time), which is significantly higher than the draft forward. We need a better one with under 300us.

lfr-0531 committed:
add _prepare_draft_tp_inputs back and some fixes.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 commented:
/bot run

lfr-0531 commented on tensorrt_llm/_torch/pyexecutor/py_executor.py:
```diff
@@ -1833,37 +1832,19 @@ def _prepare_draft_tokens(self, scheduled_requests: ScheduledRequests):
-            if spec_metadata.spec_dec_mode.is_eagle3() and hasattr(
-                    self.draft_model_engine.model.model, 'd2t'):
+            # Disable cuda graph for the 1st draft model forward
+            with self.draft_model_engine.no_cuda_graph():
```
Added a new `needs_kv_cache_recompute`.

lfr-0531 commented on tensorrt_llm/_torch/models/modeling_llama.py:
```diff
@@ -1290,6 +1290,9 @@ def forward(
         hidden_states: Optional[torch.Tensor] = None,
         **kwargs,
     ) -> torch.Tensor:
+        if hidden_states is None:
```
One model flow directly uses the `draft_model.model()`. Removed.

lfr-0531 commented on tensorrt_llm/_torch/speculative/eagle3.py:
```diff
@@ -64,66 +128,91 @@ def __post_init__(self):
+            hidden_states_read_indices, dtype=torch.long, pin_memory=True)
+        self.hidden_states_write_indices_host = torch.tensor(
+            hidden_states_write_indices, dtype=torch.long, pin_memory=True)
+        self.apply_eagle3_fc = True if is_first_draft and self.is_draft_model else False
```
Thanks for your great suggestion! I changed this part, and now it is clearer.

tensorrt-cicd commented:
[PR_Github #8690](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8690/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #8691](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8691/) [ kill ] triggered by Bot

lfr-0531 commented:
/bot run

lfr-0531 commented:
I'll test the new perf tomorrow.

lfr-0531 committed:
revert the attention metadata prepare changes.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

lfr-0531 committed:
remove prepare_draft_tp_inputs.

Signed-off-by: Fanrong Li <23290157+lfr-0531@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #8802](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8802/) [ run ] triggered by Bot

IzzyPutterman commented:
Something like this: https://github.com/NVIDIA/TensorRT-LLM/pull/5211

tensorrt-cicd commented:
[PR_Github #8907](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8907/) [ run ] triggered by Bot

lfr-0531 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #8925](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8925/) [ run ] triggered by Bot

lfr-0531 commented:
> Do we already have logic to make sure that extra pass of overlap doesnt run (and infringe on the verify pass)? I think I have an internal MR that does this from a while ago.

We didn't add it in this PR. This PR is only used to overlap the different forwards in the same iteration. After we have PR-5211, the last iteration, including the `_prepare_draft_tokens`, can be skipped.

tensorrt-cicd commented:
[PR_Github #8925](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8925/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6509](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6509/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/6509/test-results/rerun_report.html) for details.
